### DISCUSSION TITLE
Best Practices for Declaring Native Events in defineEmits

### DISCUSSION THREAD

**Discussion by alm6666** [mintysolutions/vue_snippets#22] [Q&A] [answered]
Hi everyone, I have a question regarding how native DOM events are handled in Vue’s defineEmits. 
The docs mention that if you define a native event like click inside defineEmits, it no longer listens for native DOM click events but only for component-emitted ones. 
This seems a bit confusing to me, especially if I want to listen for both native and component-emitted events. What’s the best practice here to avoid conflicts? Should Vue handle this differently?

**Comment by dgtalForest** [comment]
Great question, Adrian! 
I ran into this recently when trying to emit a custom click event from a button component while still wanting to listen for native click events on that button. I added v-on="$attrs" to pass through native events, but it felt a bit hacky. It would be nice if Vue could differentiate between native and custom events better. Right now, it’s easy to accidentally override the native event handling.

**Comment by jdnheaps** [comment]
I’ve faced the same issue! I wanted to emit a custom event and keep the native event intact, but it wasn’t clear in the documentation how to handle both. I used $listeners to make it work, but with Vue 3’s setup syntax, it’s not as intuitive. I agree with you, Emma. We need a clearer approach to manage both types of events.

**Comment by lodistack** [comment, marked as answer]
You’ve all identified a common pain point! 
The issue stems from how Vue binds custom event listeners directly to components, which can sometimes unintentionally shadow native DOM events if their names overlap. The best practice here is to use $attrs or $listeners to ensure native events are passed through, as Sam suggested. But you’re right, it can feel a bit cumbersome, especially when using defineEmits.
Vue could provide a better mechanism to handle native DOM events separately from component-emitted ones—maybe something like a dedicated v-native-on directive or a clearer option in defineEmits to distinguish the two. I’ll suggest this to the core team and see if they’re open to improving this area.

**Comment by alm6666** [comment]
Thanks! A v-native-on directive sounds like a clean solution. It would simplify the process and reduce confusion when working with both types of events.

**Comment by jdnheaps** [comment]
I agree! Having a clearer distinction between native and component-emitted events would be awesome. It would definitely prevent some of the issues I’ve run into. Thanks for the insight, Sarah!
